clippy_lints/src/methods/unnecessary_collection_clone.rs

```diff
@@ -0,0 +1,59 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::is_path_mutable;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::{deref_chain, get_inherent_method, implements_trait, make_normalized_projection};
+use rustc_errors::Applicability;
+use rustc_hir::Expr;
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty};
+use rustc_span::sym;
+
+use super::{UNNECESSARY_COLLECTION_CLONE, method_call};
+
+// FIXME: This does not check if the iter method is actually compatible with the replacement, but
+// you have to be actively evil to have an `IntoIterator` impl that returns one type and an `iter`
+// method that returns something other than references of that type.... and it is a massive
+// complicated hassle to check this
+fn has_iter_method<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    deref_chain(cx, ty).any(|ty| match ty.kind() {
+        ty::Adt(adt_def, _) => get_inherent_method(cx, adt_def.did(), sym::iter).is_some(),
+        ty::Slice(_) => true,
+        _ => false,
+    })
+}
+
+/// Check for `x.clone().into_iter()` to suggest `x.iter().cloned()`.
+//             ^^^^^^^^^ is recv
+//             ^^^^^^^^^^^^^^^^^^^^^ is expr
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, recv: &Expr<'_>) {
+    let typeck_results = cx.typeck_results();
+    let diagnostic_items = cx.tcx.all_diagnostic_items(());
+
+    // If the call before `into_iter` is `.clone()`
+    if let Some(("clone", collection_expr, [], _, _)) = method_call(recv)
+        // and the binding being cloned is not mutable
+        && let Some(false) = is_path_mutable(cx, collection_expr)
+        // and the result of `into_iter` is an Iterator
+        && let Some(&iterator_def_id) = diagnostic_items.name_to_id.get(&sym::Iterator)
+        && let expr_ty = typeck_results.expr_ty(expr)
+        && implements_trait(cx, expr_ty, iterator_def_id, &[])
+        // with an item that implements clone
+        && let Some(&clone_def_id) = diagnostic_items.name_to_id.get(&sym::Clone)
+        && let Some(item_ty) = make_normalized_projection(cx.tcx, cx.param_env, iterator_def_id, sym::Item, [expr_ty])
+        && implements_trait(cx, item_ty, clone_def_id, &[])
+        // and the type has an `iter` method
+        && has_iter_method(cx, typeck_results.expr_ty(collection_expr))
+    {
+        let mut applicability = Applicability::MachineApplicable;
```

Comment:
There is this lint trigger from Lintcheck:

> Added `clippy::unnecessary_collection_clone` at [`reqwest-0.12.5/src/dns/resolve.rs:100`](https://docs.rs/reqwest/0.12.5/src/reqwest/dns/resolve.rs.html#100)
> 
> ```
> warning: using clone on collection to own iterated items
>    --> target/lintcheck/sources/reqwest-0.12.5/src/dns/resolve.rs:100:45
>     |
> 100 |                 let addrs: Addrs = Box::new(dest.clone().into_iter());
>     |                                             ^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `dest.iter().cloned()`
>     |
>     = note: `--force-warn clippy::unnecessary-collection-clone` implied by `--force-warn clippy::all`
> ```

These cases might not be `MachineApplicable` as they change the type of the expression. To fully avoid FPs, we probably would need to track how the value is finally used, which would be hard and also prone to errors. I think this is an argument for changing the Applicability to `MaybeIncorrect` and having it in `pedantic` instead.

Reply:
Hmm, that's a shame, I hate relegating useful lints to pedantic where they won't be hit by the beginners this lint is targeting.

Would it be good if I swap the suggestion to `MaybeIncorrect` but keep in the `perf` category? Since we now have lint reasons, it would be good for a codebase to document "we don't do Iter::cloned due to XYZ" with an `allow`/`expect`.

Reply:
The applicability is only given to tools like rustfix and not directly visible to the user. So changing the applicability, shouldn't have an effect on the lint category.

The category usually depends on the FP rate and the trigger rate when we introduce the lint. If a lint is reasonable, but has too many lint triggers, it will sadly still often land in `pedantic`. We can leave the lint group discussion open for the RFC. But the applicability should still be changed.